fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,979 @@
+            self._initialized_models = True
+        # Note, if initialized_models == False, the models will still be lists
+
+        self.predictor_model = predictor_model
+        self.adversary_model = adversary_model
+
+    def _verify_dist_type(self, input_type, dist_name):
+        if callable(input_type):
+            return input_type
+        elif isinstance(input_type, str):
+            if input_type.lower() in Keyword.TYPES:
+                return input_type.lower()
+        raise ValueError(_KWARG_ERROR_MESSAGE.format(
+            dist_name, "one of {} or a callable".format(Keyword.TYPES)))
+
+    def _init_losses(self, predictor_loss, adversary_loss, predictor_function):
+        self.predictor_loss = self._verify_dist_type(predictor_loss, "predictor_loss")
+        self.adversary_loss = self._verify_dist_type(adversary_loss, "adversary_loss")
+        self.predictor_function = self._verify_dist_type(predictor_function, "predictor_function")
+
+    def _init_constraints(self, constraints):
+        """Verify the constraints and set up the corresponding network structure."""
+        self.constraints = constraints
+        if (constraints == "demographic_parity"):
+            self.pass_y = False
+        elif (constraints == "equalized_odds"):
+            self.pass_y = True
+        else:
+            raise ValueError(_KWARG_ERROR_MESSAGE.format(
+                "constraints", "\'demographic_parity\' or \'equalized_odds\'"))
+
+    def _init_cuda(self, cuda):
+        """Verify whether we can use the GPU and move pytorch model to it."""
+        if (not cuda):
+            self.cuda = False
+        elif (cuda):
+            if (not self.torch):
+                raise ValueError("Cuda is PyTorch-only")
+            if not torch.cuda.is_available():
+                raise ValueError("Cuda is not available")
+            self.cuda = True
+            self.device = torch.device("cuda:0")
+
+    def _extend_instance(self, cls_):
+        """Apply mixins to a class instance after creation."""
+        base_cls = self.__class__
+        base_cls_name = self.__class__.__name__
+        self.__class__ = type(base_cls_name, (cls_, base_cls), {})
+
+
+def _check_array(X):
+    """
+    Validate the input array, and possible coerce to 2D.
+
+    Calls :code:`sklearn.utils.check_array` on parameter X with the
+    parameters suited for Adversarial Mitigation.
+    """
+    return check_array(
+        X, accept_sparse=False, accept_large_sparse=False,
+        dtype=float, order=None, copy=False, force_all_finite=True,
+        ensure_2d=True, allow_nd=False, ensure_min_samples=1,
+        ensure_min_features=1, estimator=None
+    )
+
+
+class AdversarialMixin():
+    """The interface of a mixin class."""
+
+    def _evaluate(self, X: ndarray) -> ndarray:
+        """
+        Evaluate the model given input `X`.
+
+        Feed 2d `numpy.ndarray` through model and receive output as
+        2d `numpy.ndarray`.
+        """
+        pass
```

Comment:
according to many, and I agree with them, we shouldn't have any method which has only `pass` as its body. @riedgar-ms may disagree.